# Accessing the fly connectome dataset with CAVE

This tutorial provides a high-level overview for how to access FlyWire's dataset through CAVE. CAVE is the [connectome annotation versioning engine](https://www.biorxiv.org/content/10.1101/2023.07.26.550598v1.abstract), a service infrastructure for managing connectomics datasets and is hosted in the cloud for broad access. CAVE supports proofreading of datasets and their analysis even while proofreading is ongoing.

# CAVEclient and setup

The CAVEclient is a python library that facilitates communication with a CAVE system. It can be install with 

`pip install caveclient`

and imported like so:

In [1]:
import caveclient

## CAVE account setup

Each and every user needs to create a CAVE account and download a user token to access CAVE's services programmatically. FlyWire's data is publicly available which means that no extra permissions need to be given to a new user account to access the data.

A Google account (or Google-enabled account) is required to create a CAVE account.

#### Start here you do not have a CAVE account or are not sure

Login to CAVE to setup a new account. To do this go to this [website](https://prod.flywire-daf.com/materialize/views/datastack/flywire_fafb_public).

#### Once you have an account: Setup your token

Create a new token by running the next cell. Then, copy the token and insert it into the argument of the following cell. These two cells should be redone together to make sure that the correct token is stored on your machine. You can copy your token and store on as many machines as you like. If you think your token has been compromised just reset it but rerunning the following cell.

In [2]:
client = caveclient.CAVEclient()
client.auth.setup_token(make_new=True)

In [3]:
client.auth.save_token(token="your token goes here", overwrite=True)

## Datastacks

Datasets in CAVE are organized as datastacks. These are a combination of an EM dataset, a segmentation and a set of annotations. The datastack for FlyWire's public release is `flywire_fafb_public`. When you instantiate your client with this datastack, it loads all relevant information to access it.

In [4]:
datastack_name = "flywire_fafb_public"
client = caveclient.CAVEclient(datastack_name)

## Materialization versions

Data in CAVE is timestamped and periodically versioned - each (materialization) version cooresponds to a specific timestamp. Individual versions are made publicly available. The materialization service provides annotation queries to the dataset. It is available under `client.materialize`. 

Currently the following versions are publicly available:

In [5]:
client.materialize.get_versions()

[630]

And these are their associated timestamps:

In [6]:
for version in client.materialize.get_versions():
    print(f"Version {version}: {client.materialize.get_timestamp(version)}")

Version 630: 2023-03-21 08:10:01.194185+00:00


The client will automatically query the latest materialization version. You can specify a `materialization_version` for every query if you want to access a specific version.

# Querying the dataset

Let's have a look what annotation tables are available:

In [7]:
client.materialize.get_tables()

['fly_synapses_neuropil_v2',
 'neuron_information_v2',
 'proofread_neurons',
 'nuclei_v1',
 'synapses_nt_v1',
 'hierarchical_neuron_annotations']

## Querying neurons

The `proofread_neurons` table contains all neurons that were released in a given version. The dataset contains many more segments that were either not proofread because they are small, or belong to non-neuronal cells or other structures such as trachea. Therefore, knowing the list of all segments that represent proofread neurons is useful. It can be queried in full:

In [8]:
proofread_neurons_df = client.materialize.query_table("proofread_neurons")
proofread_neurons_df

id valid   pt_supervoxel_id          pt_root_id  \
0            1     t  78253067652813181  720575940628857210   
1            2     t  82053323167374089  720575940626838909   
2            3     t  81842766824648491  720575940626046919   
3            4     t  82405647924682371  720575940630311383   
4            5     t  82756942157675987  720575940633370649   
...        ...   ...                ...                 ...   
127973  127974     t  77198292605339502  720575940624334775   
127974  127975     t  83251928414696891  720575940625178772   
127975  127976     t  76706536096887135  720575940625567758   
127976  127977     t  75297580732428963  720575940641724661   
127977  127978     t  75649424118141806  720575940655722657   

                     pt_position  
0       [443342, 203965, 157450]  
1        [664417, 227538, 77011]  
2       [653666, 259618, 110467]  
3        [687457, 254763, 80194]  
4       [707653, 222503, 145366]  
...                          ...  
127973  [384508, 252336, 207800]  
127974  [733464, 363772, 112720]  
127975  [353200, 298616, 227080]  
127976  [273840, 204492, 246800]  
127977  [294436, 204084, 154920]  

[127978 rows x 5 columns]

In the table above each row is a proofread neuron. Segment IDs (aka neuron IDs) are called `root_ids` in CAVE. Each annotation is associated with at least one point with which data is associated; in this case `pt_root_id` and `pt_position` are the most relevant columns.

Positions in this table were calculated to be in the backbone of a neuron. This was found to be the most robust location to identify a neuron with as some do not have cell bodies and cell bodies are not central to fly neurons. There is a table that represents all _cell_ nuclei in the brain created by [Shang et al.](https://www.biorxiv.org/content/10.1101/2021.11.04.467197v1.abstract)

In [9]:
nuclei_df = client.materialize.query_table("nuclei_v1")
nuclei_df

id valid      volume   pt_supervoxel_id          pt_root_id  \
0       7393349     t   26.141245  82827379285852979  720575940626838909   
1       7416439     t   11.523400  82827998029690530  720575940627484553   
2       7415038     t   32.895959  83038623024880664  720575940626046919   
3       7415013     t   53.711176  83038760463398837  720575940630311383   
4       7415848     t    9.280717  83038554439606237  720575940633370649   
...         ...   ...         ...                ...                 ...   
143135  4389032     t  244.330332  79377386675948158  720575940629762043   
143136  8558952     t  253.359391  84870066067479558  720575940636923511   
143137  3076633     t  261.782528  78042030341189115  720575940623944072   
143138  3125634     t  274.994586  78113635431454405  720575940621426568   
143139  6655288     t  322.137088  81769305838115716  720575940629374326   

                     pt_position         bb_start_position  \
0        [709888, 227744, 57160]   [708032, 226144, 54760]   
1       [710592, 263392, 129800]  [708800, 262112, 128200]   
2        [722528, 234656, 77000]   [720768, 232896, 74480]   
3        [722912, 244032, 65200]   [720480, 242144, 62480]   
4       [721984, 229792, 119560]  [720832, 228320, 118240]   
...                          ...                       ...   
143135   [511328, 113024, 48560]   [503744, 104128, 44600]   
143136  [830720, 344608, 149840]  [821120, 335680, 146960]   
143137  [433152, 210560, 223440]  [428544, 206976, 219120]   
143138   [434208, 282112, 23400]   [429728, 278496, 19040]   
143139   [648704, 72640, 147960]   [644160, 67648, 143560]   

                 bb_end_position  
0        [711904, 229440, 59280]  
1       [712064, 264832, 131080]  
2        [724224, 236576, 79360]  
3        [725600, 246208, 67760]  
4       [723232, 231264, 120840]  
...                          ...  
143135   [520704, 122080, 53040]  
143136  [840576, 354912, 152880]  
143137  [437632, 214048, 227600]  
143138   [439264, 285568, 27320]  
143139   [653248, 77024, 152120]  

[143140 rows x 8 columns]

Not all neurons contains have a cell body in the brain (e.g. sensory, ascending) and for ~6,000 of intrinsic neurons, the segmentation did not reach up to the cell bodies which are sitting at the outer layer of the brain. In those cases, the `pt_root_id` is 0. This table also contains glia cell bodies, as well as a few false positive annotations. 

Every table has an associated description which provides further context and references to publications. This feature is provided by the annotation service which can be reached at `cave.annotation`.

In [10]:
print(client.annotation.get_table_metadata("nuclei_v1")["description"])


FlyWire nucleus description
Nucleus version: 20210322

Nuclei in this table consist of center points (in nm), volume (in μm3), and bounding boxes (in nm).

The nucleus segmentation was generated by Shang Mu (smu@princeton.edu, Seung Lab at Princeton University) using a 2D convolutional neural network (CNN) and heuristic interpolations. The training data was assembled from annotations by Selden Koolman, Merlin Moore, Sarah Morejohn, Ben Silverman, Kyle Willie, Ryan Willie, Szi-chieh Yu and Shang Mu.

As this data was generated using a 2D, rather than 3D, neural network, defects are present in the detected nuclei, particularly where there are large defects in section alignment or a number of consecutive missing sections.

False positive fragments, nucleus fragments and partial nuclei are the most common type of defects. A simple, rudimentary method for cleaning up is to disregard small fragments by thresholding by segment size or by the z-dimension of the bounding boxes. A size threshol

Positions can be useful for analysis. The CAVEclient provides some convenience functions: 

Splitting of position columns into separate x, y, and z columns

In [11]:
nuclei_df = client.materialize.query_table("nuclei_v1", split_positions=True)
nuclei_df

id valid      volume  pt_position_x  pt_position_y  \
0       7393349     t   26.141245         709888         227744   
1       7416439     t   11.523400         710592         263392   
2       7415038     t   32.895959         722528         234656   
3       7415013     t   53.711176         722912         244032   
4       7415848     t    9.280717         721984         229792   
...         ...   ...         ...            ...            ...   
143135  4389032     t  244.330332         511328         113024   
143136  8558952     t  253.359391         830720         344608   
143137  3076633     t  261.782528         433152         210560   
143138  3125634     t  274.994586         434208         282112   
143139  6655288     t  322.137088         648704          72640   

        pt_position_z   pt_supervoxel_id          pt_root_id  \
0               57160  82827379285852979  720575940626838909   
1              129800  82827998029690530  720575940627484553   
2               77000  83038623024880664  720575940626046919   
3               65200  83038760463398837  720575940630311383   
4              119560  83038554439606237  720575940633370649   
...               ...                ...                 ...   
143135          48560  79377386675948158  720575940629762043   
143136         149840  84870066067479558  720575940636923511   
143137         223440  78042030341189115  720575940623944072   
143138          23400  78113635431454405  720575940621426568   
143139         147960  81769305838115716  720575940629374326   

        bb_start_position_x  bb_start_position_y  bb_start_position_z  \
0                    708032               226144                54760   
1                    708800               262112               128200   
2                    720768               232896                74480   
3                    720480               242144                62480   
4                    720832               228320               118240   
...                     ...                  ...                  ...   
143135               503744               104128                44600   
143136               821120               335680               146960   
143137               428544               206976               219120   
143138               429728               278496                19040   
143139               644160                67648               143560   

        bb_end_position_x  bb_end_position_y  bb_end_position_z  
0                  711904             229440              59280  
1                  712064             264832             131080  
2                  724224             236576              79360  
3                  725600             246208              67760  
4                  723232             231264             120840  
...                   ...                ...                ...  
143135             520704             122080              53040  
143136             840576             354912             152880  
143137             437632             214048             227600  
143138             439264             285568              27320  
143139             653248              77024             152120  

[143140 rows x 14 columns]

Defining the position resolution. Resolutions are always in nanometers and this query asks for points to be provided in micrometers (aka 1000 nanometers):

In [12]:
nuclei_df = client.materialize.query_table("nuclei_v1", desired_resolution=[1000, 1000, 1000])
nuclei_df

id valid      volume   pt_supervoxel_id          pt_root_id  \
0       7393349     t   26.141245  82827379285852979  720575940626838909   
1       7416439     t   11.523400  82827998029690530  720575940627484553   
2       7415038     t   32.895959  83038623024880664  720575940626046919   
3       7415013     t   53.711176  83038760463398837  720575940630311383   
4       7415848     t    9.280717  83038554439606237  720575940633370649   
...         ...   ...         ...                ...                 ...   
143135  4389032     t  244.330332  79377386675948158  720575940629762043   
143136  8558952     t  253.359391  84870066067479558  720575940636923511   
143137  3076633     t  261.782528  78042030341189115  720575940623944072   
143138  3125634     t  274.994586  78113635431454405  720575940621426568   
143139  6655288     t  322.137088  81769305838115716  720575940629374326   

                       pt_position           bb_start_position  \
0        [709.888, 227.744, 57.16]   [708.032, 226.144, 54.76]   
1        [710.592, 263.392, 129.8]     [708.8, 262.112, 128.2]   
2         [722.528, 234.656, 77.0]   [720.768, 232.896, 74.48]   
3         [722.912, 244.032, 65.2]    [720.48, 242.144, 62.48]   
4       [721.984, 229.792, 119.56]   [720.832, 228.32, 118.24]   
...                            ...                         ...   
143135   [511.328, 113.024, 48.56]    [503.744, 104.128, 44.6]   
143136   [830.72, 344.608, 149.84]    [821.12, 335.68, 146.96]   
143137   [433.152, 210.56, 223.44]  [428.544, 206.976, 219.12]   
143138    [434.208, 282.112, 23.4]   [429.728, 278.496, 19.04]   
143139    [648.704, 72.64, 147.96]    [644.16, 67.648, 143.56]   

                   bb_end_position  
0         [711.904, 229.44, 59.28]  
1       [712.064, 264.832, 131.08]  
2        [724.224, 236.576, 79.36]  
3          [725.6, 246.208, 67.76]  
4       [723.232, 231.264, 120.84]  
...                            ...  
143135    [520.704, 122.08, 53.04]  
143136  [840.576, 354.912, 152.88]  
143137   [437.632, 214.048, 227.6]  
143138   [439.264, 285.568, 27.32]  
143139   [653.248, 77.024, 152.12]  

[143140 rows x 8 columns]

## Querying annotations - hierarchical annotations from Schlegel et al., 2023

[Schlegel et al](https://www.biorxiv.org/content/10.1101/2023.06.27.546055v2.abstract) introduced hierarchical annotations for all proofread neurons in the dataset. Figure 1 from their paper (shown below) outlines the hierarchy and renders individual groups of neuurons.



![](https://www.biorxiv.org/content/biorxiv/early/2023/07/15/2023.06.27.546055/F1.large.jpg)

To load all annotations (this will take ~20s):

In [13]:
hierarchical_annos_df = client.materialize.query_table("hierarchical_neuron_annotations")
hierarchical_annos_df

id_ref valid_ref   pt_supervoxel_id          pt_root_id      id valid  \
0        60947         t  83532853635373500  720575940627005443  391232     t   
1        60946         t  84307115980143621  720575940638062141  391233     t   
2        60945         t  85151060007686276  720575940634330991  391234     t   
3        60960         t  85010459891445545  720575940610409946  391235     t   
4        60961         t  83673659843320298  720575940628587324  391236     t   
...        ...       ...                ...                 ...     ...   ...   
377694  127378         t  76001267906418794  720575940640761728  760321     t   
377695  127393         t  76705299213588578  720575940644327584  760322     t   
377696  127697         t  75790161740758808  720575940637229198  760323     t   
377697  127776         t  75157049402858357  720575940627302745  760324     t   
377698  116748         t  76916955201745205  720575940639295558  760325     t   

        target_id classification_system  cell_type               pt_position  
0           60947                  flow  intrinsic  [751197, 330784, 106807]  
1           60946                  flow  intrinsic  [796396, 343997, 119089]  
2           60945                  flow  intrinsic  [845576, 316831, 145515]  
3           60960                  flow  intrinsic  [835953, 325029, 132976]  
4           60961                  flow  intrinsic  [761422, 334448, 110534]  
...           ...                   ...        ...                       ...  
377694     127378             cell_type        T4a  [314129, 204855, 184422]  
377695     127393             cell_type        T5a  [352578, 225656, 253139]  
377696     127697             cell_type        T4a  [301502, 206108, 197151]  
377697     127776             cell_type        T4a  [263197, 218814, 256978]  
377698     116748             cell_type        T5a  [367111, 257657, 246857]  

[377699 rows x 10 columns]

The `classification_system` column encodes the level of the hierarchy a given annotation belongs to. Only annotations for `flow` and `super class` are available for all neurons but finer annotations are only available for subsets of neurons. We can see that in the annotation counts for each hierarchy level:

In [14]:
hierarchical_annos_df["classification_system"].value_counts()

flow              127978
super_class       127978
cell_class         97672
cell_type          17404
cell_sub_class      6667
Name: classification_system, dtype: int64

## Subqueries - hierarchical annotations from Schlegel et al., 2023 continued

CAVE poses a limit on the size of a table that can be loaded at once. The current limit is `500,000` rows. This is to ensure the system is working for everyone and prevents accidentaly large queries to the server - that is particularly relevant for synapse queries. 

While the hierarchical annotation table is small enough to be loaded as a whole, it takes a few seconds to load. If we are only interested in a subset of the data, we can use a filter on any column to reduce the data footprint:

In [15]:
flow_annos_df = client.materialize.query_table("hierarchical_neuron_annotations", filter_equal_dict={"classification_system": "flow"})
flow_annos_df

id valid  target_id classification_system  cell_type  id_ref  \
0       391232     t      60947                  flow  intrinsic   60947   
1       391233     t      60946                  flow  intrinsic   60946   
2       391234     t      60945                  flow  intrinsic   60945   
3       391235     t      60960                  flow  intrinsic   60960   
4       391236     t      60961                  flow  intrinsic   60961   
...        ...   ...        ...                   ...        ...     ...   
127973  503938     t     109940                  flow  intrinsic  109940   
127974  503939     t     110216                  flow  intrinsic  110216   
127975  503940     t     111137                  flow  intrinsic  111137   
127976  503941     t     112845                  flow  intrinsic  112845   
127977  503942     t     127977                  flow  intrinsic  127977   

       valid_ref   pt_supervoxel_id          pt_root_id  \
0              t  83532853635373500  720575940627005443   
1              t  84307115980143621  720575940638062141   
2              t  85151060007686276  720575940634330991   
3              t  85010459891445545  720575940610409946   
4              t  83673659843320298  720575940628587324   
...          ...                ...                 ...   
127973         t  74664055675374018  720575940627905383   
127974         t  82689047382048903  720575940628019985   
127975         t  81984535373702682  720575940639551797   
127976         t  74171543118689555  720575940611095160   
127977         t  75297580732428963  720575940641724661   

                     pt_position  
0       [751197, 330784, 106807]  
1       [796396, 343997, 119089]  
2       [845576, 316831, 145515]  
3       [835953, 325029, 132976]  
4       [761422, 334448, 110534]  
...                          ...  
127973  [234729, 191865, 193858]  
127974  [702449, 370626, 173721]  
127975  [663170, 321144, 196507]  
127976  [205765, 196206, 178359]  
127977  [273840, 204492, 246800]  

[127978 rows x 10 columns]

Besides `filter_equal_dict`, the CAVEclient provides `filter_in_dict` and `filter_out_dict` as options to restrict what data is loaded. Examples:

In [16]:
cell_class_type_annos_df = client.materialize.query_table("hierarchical_neuron_annotations", filter_in_dict={"classification_system": ["cell_class", "cell_type"]})
cell_class_type_annos_df

id_ref valid_ref   pt_supervoxel_id          pt_root_id      id valid  \
0        24809         t  81350666718746524  720575940618149686  631921     t   
1         3439         t  80717416606245362  720575940628675433  631922     t   
2        42371         t  77337311972566683  720575940630166583  631923     t   
3        15670         t  80787785350455829  720575940630738683  631924     t   
4        34597         t  78676860598498552  720575940624706062  631925     t   
...        ...       ...                ...                 ...     ...   ...   
115071  127378         t  76001267906418794  720575940640761728  760321     t   
115072  127393         t  76705299213588578  720575940644327584  760322     t   
115073  127697         t  75790161740758808  720575940637229198  760323     t   
115074  127776         t  75157049402858357  720575940627302745  760324     t   
115075  116748         t  76916955201745205  720575940639295558  760325     t   

        target_id classification_system cell_type               pt_position  
0           24809            cell_class      ALIN  [624510, 288189, 127797]  
1            3439            cell_class      ALIN   [586065, 292954, 82718]  
2           42371            cell_class      ALIN  [389107, 148134, 167223]  
3           15670            cell_class      ALIN   [592115, 291905, 83257]  
4           34597            cell_class      ALIN  [468615, 298327, 130296]  
...           ...                   ...       ...                       ...  
115071     127378             cell_type       T4a  [314129, 204855, 184422]  
115072     127393             cell_type       T5a  [352578, 225656, 253139]  
115073     127697             cell_type       T4a  [301502, 206108, 197151]  
115074     127776             cell_type       T4a  [263197, 218814, 256978]  
115075     116748             cell_type       T5a  [367111, 257657, 246857]  

[115076 rows x 10 columns]

In [17]:
no_flow_annos_df = client.materialize.query_table("hierarchical_neuron_annotations", filter_out_dict={"classification_system": ["flow"]})
no_flow_annos_df

id_ref valid_ref   pt_supervoxel_id          pt_root_id      id valid  \
0        60947         t  83532853635373500  720575940627005443  391232     t   
1        60946         t  84307115980143621  720575940638062141  391233     t   
2        60945         t  85151060007686276  720575940634330991  391234     t   
3        60960         t  85010459891445545  720575940610409946  391235     t   
4        60961         t  83673659843320298  720575940628587324  391236     t   
...        ...       ...                ...                 ...     ...   ...   
377694  127378         t  76001267906418794  720575940640761728  760321     t   
377695  127393         t  76705299213588578  720575940644327584  760322     t   
377696  127697         t  75790161740758808  720575940637229198  760323     t   
377697  127776         t  75157049402858357  720575940627302745  760324     t   
377698  116748         t  76916955201745205  720575940639295558  760325     t   

        target_id classification_system  cell_type               pt_position  
0           60947                  flow  intrinsic  [751197, 330784, 106807]  
1           60946                  flow  intrinsic  [796396, 343997, 119089]  
2           60945                  flow  intrinsic  [845576, 316831, 145515]  
3           60960                  flow  intrinsic  [835953, 325029, 132976]  
4           60961                  flow  intrinsic  [761422, 334448, 110534]  
...           ...                   ...        ...                       ...  
377694     127378             cell_type        T4a  [314129, 204855, 184422]  
377695     127393             cell_type        T5a  [352578, 225656, 253139]  
377696     127697             cell_type        T4a  [301502, 206108, 197151]  
377697     127776             cell_type        T4a  [263197, 218814, 256978]  
377698     116748             cell_type        T5a  [367111, 257657, 246857]  

[377699 rows x 10 columns]

## Community annotations

## Synapse queries

### Details about FlyWire's synapses

FlyWire uses automatically annotated synapses that were produced by [Buhmann et al.](https://www.nature.com/articles/s41592-021-01183-7). Automation of synapse annotation is critical for circuit analysis but one should keep in mind that the classifier may contain biases that lead to better or worse results in different brain regions. For instance, this classifier was trained on data acquired from neuropils in the central brain and might perform worse in the optic lobe or for sensory neurons. In total, Buhmann et al. identified ~244 million _putative_ synapses.

In [18]:
client.materialize.get_annotation_count("synapses_nt_v1")

244358226

### Details about removal of false positive synapses

Each synapse is a link from a pre- to a posynaptic site. As presynapses in the fly are usually polysynaptic, there are usually multiple synapses for each presynaptic site. They assigned a `connection_score` to every synapse which can be used to filter out false positives. Buhmann et al. suggest to correlate their predictions with those from [Heinrich et al.](https://link.springer.com/chapter/10.1007/978-3-030-00934-2_36) who segmented synaptic cleft on the same dataset. This is implemented through a `cleft_score` that is associated with each synapse. We found that filtering synapses with a `cleft_score > 50` works well for removing false positives without using the `connection_score`.

Additionally, some synapses were annotated multiple times. We implemented a distance based filtering to remove redundant synapses.

## Synapse assignment to neuropils

We assigned every synapse to a neuropil based on its presynaptic location. Synapses that were outside a neuropil were assigned to the closest neuropil if that was within 10 µm. 

### TLDR: Give me the data!

We created a filtered view of the synapse table that automatically applies all established filters to remove as many false positive synapses as possible. It also adds neuropil information. For technical reasons, a different function has to be called to query this view: 

In [19]:
client.materialize.query_view("valid_synapses_nt_np", limit=10)

201 - "Limited query to 10 rows


id      pre_pt_root_id     post_pt_root_id  connection_score  cleft_score  \
0   6  720575940630786105  720575940610739662        389.255432           79   
1  11  720575940630786105  720575940545913232        469.296295          147   
2  14  720575940630786105  720575940545912208        431.531738          130   
3  15  720575940630906435  720575940627911895        475.069672          140   
4  16  720575940630906435  720575940545916816        407.200562          151   
5  18  720575940630786105  720575940545920912        217.466217          122   
6  21  720575940630786105  720575940624184764         56.978813          146   
7  22  720575940630906435  720575940545901968        478.272552          146   
8  26  720575940619149904  720575940627911895         58.005123          145   
9  27  720575940619149904  720575940545911952         17.302750          143   

       gaba       ach      glut       oct       ser        da valid_nt  \
0  0.011380  0.786194  0.003691  0.004627  0.001929  0.192179        t   
1  0.004232  0.967565  0.000087  0.000342  0.000182  0.027592        t   
2  0.000786  0.993792  0.000002  0.000006  0.000004  0.005412        t   
3  0.376253  0.278226  0.332757  0.001309  0.000762  0.010693        t   
4  0.341289  0.149622  0.499909  0.001995  0.001269  0.005916        t   
5  0.007093  0.939359  0.001130  0.000747  0.000196  0.051474        t   
6  0.006275  0.971317  0.000977  0.000983  0.000085  0.020363        t   
7  0.334301  0.341355  0.311997  0.007425  0.000349  0.004572        t   
8  0.310759  0.131895  0.522028  0.017178  0.004899  0.013241        t   
9  0.262804  0.049032  0.643314  0.018544  0.008741  0.017565        t   

   pre_pt_supervoxel_id  post_pt_supervoxel_id neuropil  \
0     76916748439646533      76916748439640439   AVLP_L   
1     76916748439639250      76916748439645298   AVLP_L   
2     76916748439651358      76916748439646534   AVLP_L   
3     76916748439657631      76916748439647565   AVLP_L   
4     76916748439651251      76916748439648883   AVLP_L   
5     76916748439663007      76916748439651348   AVLP_L   
6     76916748439651358      76916748439650144   AVLP_L   
7     76916748439651251      76916748439653785   AVLP_L   
8     76916748439651360      76916748439648983   AVLP_L   
9     76916748439651360      76916748439651344   AVLP_L   

           pre_pt_position         post_pt_position  
0  [366592, 246144, 74360]  [366584, 246216, 74320]  
1  [366560, 246144, 74440]  [366508, 246232, 74440]  
2  [366428, 246060, 74560]  [366356, 246136, 74520]  
3  [365584, 245064, 74680]  [365672, 245036, 74640]  
4  [365624, 245136, 74680]  [365712, 245200, 74720]  
5  [366252, 245948, 74640]  [366168, 245976, 74640]  
6  [366232, 246012, 74680]  [366136, 246084, 74640]  
7  [365624, 245112, 74720]  [365732, 245096, 74720]  
8  [365704, 246068, 74720]  [365656, 245948, 74720]  
9  [365756, 246080, 74720]  [365800, 245980, 74720]

This query limits the query to 10 randomly chosen rows. The maximum per query is 500,000. There are approximately 110 million synapses available in this view. Queries need to be resticted to a given set of neurons.